src/group_theory/p_group.lean

```diff
@@ -91,10 +91,27 @@ hG.of_surjective (quotient_group.mk' H) quotient.surjective_quotient_mk'
 lemma of_equiv {H : Type*} [group H] (ϕ : G ≃* H) : is_p_group p H :=
 hG.of_surjective ϕ.to_monoid_hom ϕ.surjective
 
+lemma order_of_coprime {n : ℕ} (hn : p.coprime n) (g : G) : (order_of g).coprime n :=
+let ⟨k, hk⟩ := hG g in (hn.pow_left k).coprime_dvd_left (order_of_dvd_of_pow_eq_one hk)
+
+/-- If `gcd(p,n) = 1`, then the `n`th power map is a bijection. -/
+noncomputable def pow_equiv' {n : ℕ} (hn : p.coprime n) : G ≃ G :=
+let h : ∀ g : G, (nat.card (subgroup.zpowers g)).coprime n :=
+λ g, order_eq_card_zpowers' g ▸ hG.order_of_coprime hn g in
+{ to_fun := (^ n),
+  inv_fun := λ g, (pow_coprime (h g)).symm ⟨g, subgroup.mem_zpowers g⟩,
+  left_inv := λ g, subtype.ext_iff.1 $ (pow_coprime (h (g ^ n))).left_inv
+    ⟨g, _, subtype.ext_iff.1 $ (pow_coprime (h g)).left_inv ⟨g, subgroup.mem_zpowers g⟩⟩,
+  right_inv := λ g, subtype.ext_iff.1 $ (pow_coprime (h g)).right_inv ⟨g, subgroup.mem_zpowers g⟩ }
```

Comment:
```suggestion
/-- If `gcd(p,n) = 1`, then the `n`th power map is a bijection. -/
noncomputable def pow_equiv' {n : ℕ} (hn : p.coprime n) : G ≃ G :=
let h : ∀ g : G, (nat.card (subgroup.zpowers g)).coprime n :=
  λ g, order_eq_card_zpowers' g ▸ hG.order_of_coprime hn g in
{ to_fun := (^ n),
  inv_fun := λ g, (pow_coprime (h g)).symm ⟨g, subgroup.mem_zpowers g⟩,
  left_inv := λ g, subtype.ext_iff.1 $ (pow_coprime (h (g ^ n))).left_inv
    ⟨g, _, subtype.ext_iff.1 $ (pow_coprime (h g)).left_inv ⟨g, subgroup.mem_zpowers g⟩⟩,
  right_inv := λ g, subtype.ext_iff.1 $ (pow_coprime (h g)).right_inv ⟨g, subgroup.mem_zpowers g⟩ }
```

Thanks, this is a clever solution! My approach is indeed more complicated than I thought and I need another lemma to make it work (which should be added to *order_of_elements* IMO):
```
theorem order_of_pow_coprime {G} [monoid G] {g : G} {n : ℕ} (h : (order_of g).coprime n) :
  order_of (g ^ n) = order_of g :=
begin
  obtain rfl | hn := eq_or_ne n 1, { rw pow_one },
  rw [order_of_pow'', h.gcd_eq_one, nat.div_one],
  rw [← order_of_pos_iff, pos_iff_ne_zero],
  intro h0, rw [h0, n.coprime_zero_left] at h,
  exact hn h,
end
```
and then 
```
noncomputable def pow_equiv' {n : ℕ} (hn : p.coprime n) : G ≃ G :=
{ to_fun := (^ n),
  inv_fun := λ g, g ^ ((order_of g).gcd_b n),
  left_inv := λ g, begin
    dsimp only, have := hG.order_of_coprime hn g,
    rw [order_of_pow_coprime this, ←zpow_coe_nat, ←zpow_mul, zpow_eq_mod_order_of,
      mul_gcd_b_mod, ←zpow_eq_mod_order_of, zpow_coe_nat, this.gcd_eq_one, pow_one],
  end,
  right_inv := λ g, begin
    dsimp only, rw [←zpow_coe_nat, ←zpow_mul, mul_comm, zpow_eq_mod_order_of, mul_gcd_b_mod,
      ←zpow_eq_mod_order_of, zpow_coe_nat, (hG.order_of_coprime hn g).gcd_eq_one, pow_one],
  end }
```

However, your approach is more succinct and the reuse of `pow_coprime` is very nice. So I suggest that you just add a lemma:
```
lemma pow_equiv'_symm_apply {n : ℕ} (hn : p.coprime n) (g : G) :
  (hG.pow_equiv' hn).symm g = g ^ ((order_of g).gcd_b n) :=
by { rw order_eq_card_zpowers', refl }
```

Reply:
#17252